### PR TITLE
Extend containerrun helper with UDS command channel

### PR DESCRIPTION
**Pull request by andreas-kupries**
## Description

Containerrun is modified to listen on a container local UDS for commands to stop/start the processes it controls. Enables temporary suspensions for (human) operator work without making it visible to higher kube layers

## Motivation and Context

  * https://jira.suse.com/browse/CAP-1305

## How Has This Been Tested?

Additional unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by andreas-kupries** [comment]
Certainly. And apologies for not thinking of the docs at all from the beginning.

File: container-run/pkg/containerrun/containerrun.go [hunk 1]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+}
+
+func handlePacket(
+	conn PacketConnection,
+	commands chan processCommand,
+) {
+	defer conn.Close()
+
+	packet := make([]byte, 256)
+	count, _, _ := conn.ReadFrom(packet)
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Please read the referenced https://johnrefior.com/gobits/read?post=12

File: container-run/pkg/containerrun/containerrun.go [hunk 2]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+) error {
+	err := startMainProcess (
+		runner,
+		command,
+		stdio,
+		processRegistry,
+		errors,
+		done)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Do you really want
```
     <string>: failed to start processes: failed to run container: ¯\\_(ツ)_/¯
```
over
```
      <string>: failed to run container: ¯\\_(ツ)_/¯
```

File: container-run/pkg/containerrun/containerrun.go [hunk 3]
```diff
@@ -54,30 +66,233 @@ func Run(
...
 	if err != nil {
 		return &runErr{err}
 	}
 	processRegistry.Register(process)
 
-	if postStartCommandName != "" {
-		if commandChecker.Check(postStartCommandName) {
+	go func() {
+		if err := process.Wait(); err != nil {
+			errors <- err
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: No. This way you get `failed to run container: failed to run container: ...`

File: container-run/pkg/containerrun/containerrun.go [hunk 4]
```diff
@@ -360,3 +555,36 @@ func (cc *CommandChecker) Check(command string) bool {
...
+type PacketListener interface {
+	ListenPacket(network, address string) (PacketConnection, error)
+}
+
+// NetPacketListener satisfies the PacketListener interface.
+type NetPacketListener struct {
+	listen func(network, address string) (net.PacketConn, error)
+}
+
+// NewNetPacketListener constructs a new NetPacketListener
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 5]
```diff
@@ -360,3 +555,36 @@ func (cc *CommandChecker) Check(command string) bool {
...
+func NewNetPacketListener(
+	listen func(network, address string) (net.PacketConn, error),
+) *NetPacketListener {
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 6]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+
+	for {
+		select {
+		case cmd := <-commands:
+			// Note: Commands are ignored if the system is
+			// already in the requested state. IOW
+			// demanding things to stop when things are
+			// already stopped does nothing. Similarly for
+			// demanding a start when the children are
+			// started/up/active.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 7]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+			packet, err := listener.ListenPacket("unixgram", sockAddr)
+			if err != nil {
+				errors <- err
+			}
+			handlePacket(packet, commands)
+		}
+	}()
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Not taken.

File: container-run/pkg/containerrun/containerrun.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Not taken.

File: container-run/pkg/containerrun/containerrun.go [hunk 8]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+	if err != nil {
+		return err
+	}
+
+	go processRegistry.HandleSignals(sigs, errors)
+
+	active := true
+
+	err = watchForCommands (listener, socketToWatch, errors, commands)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 9]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+		processRegistry,
+		errors,
+		done)
+	if err != nil {
+		return err
+	}
+
+	go processRegistry.HandleSignals(sigs, errors)
+
+	active := true
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 10]
```diff
@@ -360,3 +555,36 @@ func (cc *CommandChecker) Check(command string) bool {
...
+	return &NetPacketListener{
+		listen: listen,
+	}
+}
+
+// ListenPacket implements listening for packets.
+func (npl *NetPacketListener) ListenPacket(network, address string) (PacketConnection, error) {
+	conn, error := npl.listen(network, address)
+	return conn, error
+}
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit 725a12fbe3e2ae6038708b4ba82bdd2b34d3c774 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go [hunk 11]
```diff
@@ -54,30 +66,233 @@ func Run(
...
+	switch command {
+	case ProcessStart, ProcessStop:
+		commands <- processCommand (command)
+	default:
+		// Bad commands are ignored. Else they could be used to DOS the runner.
+	}
+}
+
+func stopProcesses (processRegistry *ProcessRegistry) {
+	_ = processRegistry.SignalAll(os.Kill)
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Hm. Failing to kill the children. Ok, yes. That is something the operator has to know, that the processes are still present and interfering. Of course, this end `containerrun`, the container, and thus the processes the hard way.

Done with commit a43bc87a216f8433cf38f3ee4c6b64ee891881e7 (not yet pushed).

File: container-run/pkg/containerrun/containerrun.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Done with commit d8de0c2ae2b242230deccdc77b11832a9a634fbf.

File: container-run/pkg/containerrun/containerrun.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: Given the reorg of the runErr wrapping ... Done with commit d8de0c2ae2b242230deccdc77b11832a9a634fbf.

File: container-run/pkg/containerrun/containerrun.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply container-run/pkg/containerrun/containerrun.go]: given the reorg of the runErr wrapping ... Done with commit d8de0c2ae2b242230deccdc77b11832a9a634fbf.

**Comment by andreas-kupries** [comment]
@f0rmiga Pushed reworked containerrun helper and changes to factory. Passes vet, lint, unit tests. Cluster starts ok with the modified operator.
(Today was pretty all just fixing the issues in the redo which prevented that).

Still has issues when actually used. 
While it reacts to `STOP` commands it __kills the targeted container__ somehow, still. 
I tried this using the `nats` pod.
Self-killed `nats-nats` once,
and killed the `loggregator-agent-loggregator-agent` from the `nats-nats` container through the visible socket on second attempt.
Find it weird that the `nc ...` command is not exiting after sending the packet.

I currently have no log data telling me what is going on.
Might be looking in the wrong place (kube container logs).

**Comment by andreas-kupries** [comment]
Wrt side note: It seems by default `nc` will listen for a response, even for datagrams. As the helper does not send any `nc` waits forever. Solution is `-w 1` to set a single-second timeout.

Using `stern` does not really help ... Looking just at the `nats` pod the operation generates no logs at all in all of the containers. ... And looking at the entire system swamps whatever signal there might be in noise from everything else.

It seems that adding log output to `containerrun` itself might be the best bet for investigating this.

Right now the only idea I have is that `containerrun` sends the kill signal not just to its children, but to itself as well. Oh! That or the children going away is treated as some kind of error (it should not react to sigchld and such, there is code to prevent that).

**Comment by andreas-kupries** [comment]
@f0rmiga @viovanov

With the logging it became clear that my assumption on what `Wait` would return, and how to handle the datagram listener were wrong.

So, this looks to be working as intended now, STOP and START cause `containerun` to kill and restart the job processes.

We are still fighting the system however.

Stop the `blobstore`, for example. Then wait for kube to realize that the job is not properly responding, through its probes (live, readiness, whichever), and see it restarting the job container. Very unwanted.

To the system our removal of the job process is nothing more than a problem in the component it has to solve, and then does, by creating a working component from scratch.

---

I am beginning to think that we are on the wrong path here.
As it is EOD I will defer discussion of other possible paths for migration to tomorrow.
That said, I currently have no proper ideas on how to work with the system, instead of against it.

Except in general to look at the parts where we have to restore something and then see if the running component can be convinced to perform the restoration for us in some way.

The old instructions do that in part by going through `mysql` and `mysqld` for the CCDB. In part only because it still assumes that the CC and anything else talking to the database will be (made to be) quiet while the new database gets uploaded.

File: container-run/pkg/containerrun/containerrun.go [hunk 12]
```diff
@@ -39,13 +51,21 @@ func Run(
...
 	postStartCommandName string,
 	postStartCommandArgs []string,
 	postStartConditionCommandName string,
 	postStartConditionCommandArgs []string,
 ) error {
+	fmt.Printf ("containerrun: %s / %s\n", jobName, processName)
+	fmt.Printf ("         cmd: %v\n", args)
+	fmt.Printf ("        post: %s %v\n", postStartCommandName, postStartCommandArgs)
+	fmt.Printf ("   post cond: %s %v\n", postStartConditionCommandName, postStartConditionCommandArgs)
+
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]: We need to get rid of debugging statements.

```suggestion
```

File: container-run/pkg/containerrun/containerrun.go [hunk 13]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	}
+
+	go processRegistry.HandleSignals(sigs, errors)
+
+	// This flag records the state of the system and its child
+	// processes. It is set to true when the child processes are
+	// running, and false otherwise.
+	active := true
+
+	if err = watchForCommands (listener, jobName, processName, errors, commands); err != nil {
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	if err := watchForCommands (listener, jobName, processName, errors, commands); err != nil {
```

File: container-run/pkg/containerrun/containerrun.go [hunk 14]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+		return err
+	}
+
+	fmt.Println ("Processing activities ...")
+
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
```

File: container-run/pkg/containerrun/containerrun.go [hunk 15]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+		stdio,
+		command,
+		postStartCommand,
+		conditionCommand,
+		processRegistry,
+		errors,
+		done)
+	if err != nil {
+		return err
+	}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	if err := startProcesses (
		runner,
		conditionRunner,
		commandChecker,
		stdio,
		command,
		postStartCommand,
		conditionCommand,
		processRegistry,
		errors,
		done,
	); err != nil {
		return err
	}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 16]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+						stdio,
+						command,
+						postStartCommand,
+						conditionCommand,
+						processRegistry,
+						errors,
+						done)
+					if err != nil {
+						return err
+					}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
					if err := startProcesses (
						runner,
						conditionRunner,
						commandChecker,
						stdio,
						command,
						postStartCommand,
						conditionCommand,
						processRegistry,
						errors,
						done,
					); err != nil {
						return err
					}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 17]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+
+					active = true
+				} else {
+					fmt.Println ("Processing START. Ignored (system running).")
+				}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
				}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 18]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+			}
+		case <-done:
+			// Ignore a done process when we actively
+			// stopped the children via ProcessStop.
+			if (active) {
+				fmt.Println ("Processing DONE. Ending container.")
+				return nil
+			}
+			fmt.Println ("Processing DONE. Ignored (system is stopped).")
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
			if active {
				return nil
			}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 19]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+		case err := <-errors:
+			// Ignore done signals when we actively
+			// stopped the children via ProcessStop.
+			// Wait returns with !state.Sucess, `signal: killed`
+			if (active) {
+				fmt.Printf ("Processing ERR: %s. Ending container", err)
+				return err
+			}
+			fmt.Printf ("Processing ERR: %s. Ignored (system is stopped)\n", err)
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
			if active {
				return err
			}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 20]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	errors chan error,
+	commands chan processCommand,
+) error {
+	sockAddr := fmt.Sprintf ("/var/vcap/data/%s/%s_containerrun.sock", jobName, processName)
+
+	go func() {
+		for {
+			fmt.Printf ("Waiting for commands at %s\n", sockAddr)
+
+			if err := os.RemoveAll(sockAddr); err != nil {
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
			if err := os.RemoveAll(sockAddr); err != nil {
```

File: container-run/pkg/containerrun/containerrun.go [hunk 21]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	command := strings.TrimSpace(string(packet[:n]))
+	fmt.Printf ("Received command `%s`.\n", command)
+
+	switch command {
+	case ProcessStart, ProcessStop:
+		commands <- processCommand (command)
+	default:
+		// Bad commands are ignored. Else they could be used to DOS the runner.
+		fmt.Println ("Bogus command ignored.")
+	}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	default:
		// Bad commands are ignored. Else they could be used to DOS the runner.
	}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 22]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+
+	packet := make([]byte, 256)
+	n, _, err := conn.ReadFrom(packet)
+	// Return address ignored. We do not send anything out.
+	if err != nil && err != io.EOF {
+		errors <- fmt.Errorf("failed to read command: %v", err)
+	}
+
+	command := strings.TrimSpace(string(packet[:n]))
+	fmt.Printf ("Received command `%s`.\n", command)
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	command := strings.TrimSpace(string(packet[:n]))
```

File: container-run/pkg/containerrun/containerrun.go [hunk 23]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+		fmt.Println ("Bogus command ignored.")
+	}
+}
+
+func stopProcesses (processRegistry *ProcessRegistry, errors chan<- error) {
+	fmt.Println ("Stopping processes ...")
+	for _, err := range processRegistry.SignalAll(os.Kill) {
+		errors <- err
+	}
+}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
func stopProcesses (processRegistry *ProcessRegistry, errors chan<- error) {
	for _, err := range processRegistry.SignalAll(os.Kill) {
		errors <- err
	}
}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 24]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	stdio Stdio,
+	command Command,
+	postStartCommand Command,
+	conditionCommand Command,
+	processRegistry *ProcessRegistry,
+	errors chan error,
+	done chan struct{},
+) error {
+	fmt.Println ("Starting processes ...")
+
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
) error {
```

File: container-run/pkg/containerrun/containerrun.go [hunk 25]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	err := startMainProcess (
+		runner,
+		command,
+		stdio,
+		processRegistry,
+		errors,
+		done)
+	if err != nil {
+		return err
+	}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	if err := startMainProcess (
		runner,
		command,
		stdio,
		processRegistry,
		errors,
		done,
	); err != nil {
		return err
	}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 26]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+func startMainProcess (
+	runner Runner,
+	command Command,
+	stdio Stdio,
+	processRegistry *ProcessRegistry,
+	errors chan error,
+	done chan struct{},
+) error {
+	fmt.Println ("Starting main process ...")
+
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
) error {
```

File: container-run/pkg/containerrun/containerrun.go [hunk 27]
```diff
@@ -54,75 +74,314 @@ func Run(
...
-		if commandChecker.Check(postStartCommandName) {
+	go func() {
+		fmt.Println ("Waiting for main process to complete ...")
+		if err := process.Wait(); err != nil {
+			errors <- &runErr{err}
+			return
+		}
+		fmt.Println ("Main process is done.")
+		done <- struct{}{}
+	}()
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
	go func() {
		if err := process.Wait(); err != nil {
			errors <- &runErr{err}
			return
		}
		done <- struct{}{}
	}()
```

File: container-run/pkg/containerrun/containerrun.go [hunk 28]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+	postStartCommand Command,
+	conditionCommand Command,
+	processRegistry *ProcessRegistry,
+	errors chan error,
+) {
+	if postStartCommand.Name != "" {
+		if commandChecker.Check(postStartCommand.Name) {
 			go func() {
+				fmt.Printf ("Post-start timeout: %v\n", postStartTimeout)
+
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
```

File: container-run/pkg/containerrun/containerrun.go [hunk 29]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+			// already in the requested state. I.e
+			// demanding things to stop when things are
+			// already stopped does nothing. Similarly for
+			// demanding a start when the children are
+			// started/up/active.
+
+			switch cmd {
+			case ProcessStop:
+				if active {
+					fmt.Println ("Processing STOP. Killing childs.")
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
				if active {
```

File: container-run/pkg/containerrun/containerrun.go [hunk 30]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+					// To properly ignore these
+					// events the flag has to be
+					// set before sending any
+					// signals.
+
+					active = false
+					stopProcesses(processRegistry, errors)
+				} else {
+					fmt.Println ("Processing STOP. Ignored (system is stopped).")
+				}
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
				}
```

File: container-run/pkg/containerrun/containerrun.go [hunk 31]
```diff
@@ -54,75 +74,314 @@ func Run(
...
+			case ProcessStart:
+				if !active {
+					fmt.Println ("Processing START. Restarting children.")
```
Comment (f0rmiga) [inline container-run/pkg/containerrun/containerrun.go]:
```suggestion
				if !active {
```

**Pull request merged**